src/coreclr/jit/utils.cpp

```diff
@@ -2456,6 +2455,42 @@ float FloatingPointUtils::minimum(float val1, float val2)
     return isNegative(val1) ? val1 : val2;
 }
 
+//------------------------------------------------------------------------
+// normalize: Normalize a floating point value.
+//
+// Arguments:
+//    value - the value
+//
+// Return Value:
+//    Normalized value.
+//
+// Remarks:
+//   This is a no-op on all host platforms but x86. On x86 floats are returned on
+//   the x87 stack. Since `fld` will automatically quiet signalling NaNs this
+//   means that it is very easy for a float to nondeterministically change bit
+//   representation if it is a snan, depending on whether a function that
+//   returns the value is inlined or not by the C++ compiler. To get around the
+//   nondeterminism we quiet the NaNs ahead of time as a best-effort fix.
+//
+double FloatingPointUtils::normalize(double value)
```

Comment:
This seems reasonable to me and is should be within the allowed handling by IEEE 754.

----------------

The potential downside is that some advanced user is trying to utilize such constants with an advanced SIMD algorithm and the relevant floating-point bitwise manipulation instructions (`andps`, `orps`, `xorps`, etc).

I expect the most likely bitwise values used will be:
* AllBitsSet - A qNaN
* NegativeZero - Only sign bit set
* PositiveInfinity - All exponent bits set
* All mantissa bits set - A subnormal number

None of these should be prone to complications from this normalization. `All mantissa bits set` has a potential separate issue where it "can" be subject to an FPU mode where subnormal values are normalized to Zero. However, we don't support this and none of our current target platforms do this except for Arm32 where it doesn't have a relevant control switch for SIMD (scalar logic is not subject to the same issue).

----------------

Unfortunately we cannot do something clever like utilize the relevant MMX instructions to bypass `FLD`. This is because while `mm0` maps to `FP(0)`, the MMX registers only expose the lower 64-bits and set the upper 16-bits to all ones, there-by changing the value that a x87 FPU instruction would see.

We could, however, change the managed calling convention to just not use the x87 FPU entirely. Then this value normalization would only be a consideration when interoping with native code. That may also be faster and simplify some of the overall handling given that we don't use the x87 FPU for anything other than returning floating-point values on x86.

Reply:
> We could, however, change the managed calling convention to just not use the x87 FPU entirely. Then this value normalization would only be a consideration when interoping with native code. That may also be faster and simplify some of the overall handling given that we don't use the x87 FPU for anything other than returning floating-point values on x86.

The original problem here is in the JIT's code generated by MSVC. We get non-determinism because the code produced by MSVC sometimes will quiet the NaNs and sometimes won't, depending on inlining/link-time code generation.

Changing the managed calling convention here is a bit more work than I'd like to take on at this point, but I agree that it is likely user's can hit the same behavior with the JIT that we are hitting with MSVC, depending on inlining.